pkg/volume/util/util.go

```diff
@@ -270,3 +270,119 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+	}
+	glog.Errorf("#### DEBUG LOG ####: mapDevice devicePath: %s", devicePath)
+	glog.Errorf("#### DEBUG LOG ####: mapDevice mapDir; %s", mapDir)
+
+	_, err := os.Stat(mapDir)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("cannot validate map path: %s", mapDir)
+		return err
+	}
+	if err = os.MkdirAll(mapDir, 0750); err != nil {
+		return fmt.Errorf("Failed to mkdir %s, error", mapDir)
+	}
+	// If old symlink exits, remove it then create new one
+	linkPath := mapDir + "/" + string(linkName)
+	if islinkExist, checkErr := IsSymlinkExist(linkPath); checkErr != nil {
+		return checkErr
+	} else if islinkExist {
+		if err = os.Remove(linkPath); err != nil {
+			return err
+		}
+	}
+	if err := os.Symlink(devicePath, linkPath); err != nil {
+		return err
+	}
+	glog.Errorf("#### DEBUG LOG ####: mapDevice: no error")
+	return nil
+}
+
+// UnmapDevice removes a symbolic link to block device under specified map path
+func UnmapDevice(mapDir string, linkName string) error {
+	if len(mapDir) == 0 {
+		return fmt.Errorf("Failed to unmap device from map path. mapDir is empty")
+	}
+	glog.Errorf("#### DEBUG LOG ####: unmapDevice mapDir; %s", mapDir)
+
+	// Check symbolic link exists
+	linkPath := mapDir + "/" + string(linkName)
+	if islinkExist, checkErr := IsSymlinkExist(linkPath); checkErr != nil {
+		return checkErr
+	} else if !islinkExist {
+		glog.Warningf("Warning: Unmap skipped because symlink does not exist on the path: %v", linkPath)
+		return nil
+	}
+	if err := os.Remove(linkPath); err != nil {
+		return err
+	}
+	glog.Errorf("#### DEBUG LOG ####: unmapDevice find symlink")
+	return nil
+}
+
+// RemovePath removes a file or directory on specified map path
+func RemoveMapPath(mapPath string) error {
+	if len(mapPath) == 0 {
+		return fmt.Errorf("Failed to remove map path. mapPath is empty")
+	}
+	glog.Errorf("#### DEBUG LOG ####: RemoveGlobalMapPath mapPath; %s", mapPath)
+	err := os.Remove(mapPath)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	glog.Errorf("#### DEBUG LOG ####: RemoveMapPath succeed")
+	return nil
+}
+
+// IsSymlinkExist retruns true if specified symbolic link exists
+func IsSymlinkExist(mapPath string) (bool, error) {
+	fi, err := os.Lstat(mapPath)
+	if err == nil && fi.Mode()&os.ModeSymlink != 0 {
+		return true, nil
+	} else if os.IsNotExist(err) {
+		return false, nil
+	} else {
+		return false, err
+	}
+}
+
+// GetDeviceSymlinkRefs searches symbolic links under global map path
+func GetDeviceSymlinkRefs(devPath string, mapPath string) ([]string, error) {
+	var refs []string
+	fis, err := ioutil.ReadDir(mapPath)
+	if err != nil {
+		return nil, fmt.Errorf("Directory cannot read %v", err)
+	}
+	for _, fi := range fis {
+		if fi.Mode()&os.ModeSymlink != os.ModeSymlink {
+			continue
+		}
+		filename := fi.Name()
+		path, err := os.Readlink(mapPath + "/" + filename)
+		if err != nil {
+			return nil, fmt.Errorf("Symbolic link cannot be retrieved %v", err)
+		}
+		if strings.Contains(path, devPath) {
```

Comment:
Should be the strings equal?

What if the symlink points to another device? It's used to scan global directory of a volume and the volume should have only one device. As I understand it, something went terribly wrong if there are mixed symlinks to different devices in one global dir. Should we remove the stale ones?

Reply:
So the big difference between mount and symlink, mount point is always equal to corresponding block device, but symbolic link might refer wrong block device.

(ex, after rebooting node, order of device might be changed. Kubelet recreate  symlink during start up, therefore I suppose this should not happen.)

Do you think we should not obtain physical device from destination of symbolic link?
If so, even if destination of symbolic link is wrong/corrupted, we don't detach unexpected volume. Then symlink is used for just counting number.

>> Should be the strings equal?
Agree. To avoid unexpected selecting a volume, this should check that the strings are equal.

Here is example of Global map path. There are multiple symlinks but destination of them are same.
```
Global map path for "Block" volumeMode volume
/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/{pod uuid1}
/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/{pod uuid2}
...
```

>> Should we remove the stale ones?

We may need this logic at somewhere. We can check existence of  pod using UUID.

Reply:
Can we require (or strongly recommend) that WaitForAttach() and SetupUpDevice() returns device name that's stable across reboots? For most devices it would be something in `/dev/disk/by-id`, I am not sure about all FC/iSCSI/multipath devices though.

> Do you think we should not obtain physical device from destination of symbolic link?

I don't know. What happens after reboot? Will kubelet run `SetUpDevice()` again and re-create the symlinks or will it assume that the device is already set up and it will just re-use the links (even though they may point to an old device)?

> If so, even if destination of symbolic link is wrong/corrupted, we don't detach unexpected volume. Then symlink is used for just counting number.

Yeah, links for reference counting is OK.

Reply:
Requiring would be nice if we can enforce it (I don't see how).  Otherwise we have to assume the plugin could mess up.  And it's not just reboots, also adding/removing devices while the node is still running can cause issues if you don't use a stable path.  Perhaps, one volume at /dev/sdb could be removed, and a different volume (even different volume type) could be added at /dev/sdb afterwards.  Can we detect and protect against this?

On a kubelet restart, there's a few scenarios we have to consider when detecting existing links and whether or not we can trust what it's pointing to:
1. Pod is still running on the node: The pod will be added to dsw, and in the reconciler, it will go through setup, then update asw.  This should be safe because we have the full volume spec and can override stale links.
2. Pod is deleted: Volume reconstruction scans pod directories for volume links and adds the volume to asw.  Can we trust the links here?  Only if volume plugin implements reconstruction correctly?  What is the danger if the links are stale?

Reply:
I'd add getDevicePath() into BlockVolumeMapper interface to obtain stable path.
We require that the getDevicePath() return path of symlink under /dev/disk/by-id/, and the check return value if it's contains /dev/disk/by-id path.

>>Pod is still running on the node: The pod will be added to dsw, and in the reconciler, it will go through setup, then >>update asw. This should be safe because we have the full volume spec and can override stale links.

Agree.

>>Pod is deleted: Volume reconstruction scans pod directories for volume links and adds the volume to asw. Can we >>trust the links here? Only if volume plugin implements reconstruction correctly? What is the danger if the links are stale?

I think we shouldn't trust the destination of link. Should we store full name and path of /dev/disk/by-id/{link} on anywhere? If we have this, we can refer correct volume without volume spec.

>>also adding/removing devices while the node is still running can cause issues if you don't use a stable path. 
>>Perhaps, one volume at /dev/sdb could be removed, and a different volume (even different volume type) could be >> added at /dev/sdb afterwards. Can we detect and protect against this?

I confirmed that when we use block device at docker, we can't detect even if the attached device is silently replaced.
```
% sudo docker run --device=/dev/loop2:/dev/xvdc --rm -it centos bash
[root@9ad607444fe9 /]# md5sum /dev/xvdc
53c0a6941d5f08024fda66e2ae3c934d  /dev/xvdc

- After detach and replace backend file of /dev/loop2 from fcfile to fcfile2. Only checksum is different.

[root@9ad607444fe9 /]# md5sum /dev/xvdc
01732c77d962867ce7bd125830634a53  /dev/xvdc

- If device is gone, just fail to writing operation with No space left on device.
[root@9ad607444fe9 /]# dd if=/dev/zero of=/dev/xvdc
dd: writing to '/dev/xvdc': No space left on device
1+0 records in
0+0 records out
0 bytes (0 B) copied, 0.0178061 s, 0.0 kB/s
```

Reply:
The Setup() already should return the stable path.  Where else do we need to get the stable path?

I don't think we can enforce "/dev/disk/by-id" path.  On some OSes, we may not have permission to modify it.

Reply:
@mtanino can you also try this sequence?

1. Pass in unique symlink to docker, ie /dev/unique1 -> /dev/loop2
2. Remove the device
3. Add a new device with a different unique symlink, but points to the same device path, ie /dev/unique2 -> /dev/loop2

Is the container able to access the new device?

Reply:
@msau42 
I taleked @msau42 directly, even if we follow above steps, container had access to newly added /dev/loop2. Therefore we guess container itself has resolved device path inside even if we pass symlink at first.

Reply:
@msau42 @jsafrane

As for the volume reconstruction case 2(pod is deleted), I think we can believe newly created volumeSpec during ConstructVolumeSpec at each plugin.
I checked several volume plugins, they create new volume spec from only information of mountpath. Even if we introduce block volume support, I think we can use same way because newly introduced `global map path` still contain volume plugins unique information like this.

- ./kubernetes.io/fc/volumeDevices/`28000001ff0414e2-lun-0`/8576273a-ac4f-11e7-bc66-525400f59db4

But I think we need separated volume construction method like ConstructVolumeSpecForBlock() since path is different from filesystem case.

plugin | way to create new volume spec
-- | --
aws | get VolumeID from mountpoint path on plugin directory and create new volumespec from it
azure_dd | get DataDiskURI from mountpoint path on plugin directory
gce_pd | get PDName from mountpoint path on plugin directory
portworxglusterstorage os | create a new volumespec from volumename only(maybe volumename is unique key?)